pkg/controller/datavolume-controller.go

```diff
@@ -438,88 +438,92 @@ func (r *DatavolumeReconciler) Reconcile(_ context.Context, req reconcile.Reques
+				snapshotClassName, _ := r.getSnapshotClassForSmartClone(datavolume, pvcSpec)
+				return r.reconcileSmartClonePvc(log, datavolume, pvcSpec, transferName, snapshotClassName)
+			}
+			if selectedCloneStrategy == CsiClone {
+				csiDriverAvailable, err := r.storageClassCSIDriverExists(pvcSpec.StorageClassName)
+				if err != nil && !k8serrors.IsNotFound(err) {
+					return reconcile.Result{}, err
+				}
+				if !csiDriverAvailable {
+					// err csi clone not possible
+					return reconcile.Result{},
+						r.updateDataVolumeStatusPhaseWithEvent(cdiv1.CloneScheduled, datavolume, pvc, selectedCloneStrategy,
+							DataVolumeEvent{
+								eventType: corev1.EventTypeWarning,
+								reason:    ErrUnableToClone,
+								message:   fmt.Sprintf("CSI Clone configured, but no CSIDriver available for %s", *pvcSpec.StorageClassName),
+							})
+				}
 
-		newPvc, err := r.createPvcForDatavolume(log, datavolume, pvcSpec)
-		if err != nil {
-			if errQuotaExceeded(err) {
-				r.updateDataVolumeStatusPhaseWithEvent(cdiv1.Pending, datavolume, nil, selectedCloneStrategy,
-					DataVolumeEvent{
-						eventType: corev1.EventTypeWarning,
-						reason:    ErrExceededQuota,
-						message:   err.Error(),
-					})
+				return r.reconcileCsiClonePvc(log, datavolume, pvcSpec, transferName)
 			}
-			return reconcile.Result{}, err
-		}
-		pvc = newPvc
-	} else {
-		if getSource(pvc) == SourceVDDK {
-			changed, err := r.getVddkAnnotations(datavolume, pvc)
+
+			newPvc, err := r.createPvcForDatavolume(log, datavolume, pvcSpec)
 			if err != nil {
+				if errQuotaExceeded(err) {
+					r.updateDataVolumeStatusPhaseWithEvent(cdiv1.Pending, datavolume, nil, selectedCloneStrategy,
+						DataVolumeEvent{
+							eventType: corev1.EventTypeWarning,
+							reason:    ErrExceededQuota,
+							message:   err.Error(),
+						})
+				}
 				return reconcile.Result{}, err
 			}
-			if changed {
-				err = r.client.Get(context.TODO(), req.NamespacedName, datavolume)
+			pvc = newPvc
+		} else {
+			if getSource(pvc) == SourceVDDK {
+				changed, err := r.getVddkAnnotations(datavolume, pvc)
 				if err != nil {
 					return reconcile.Result{}, err
 				}
+				if changed {
+					err = r.client.Get(context.TODO(), req.NamespacedName, datavolume)
+					if err != nil {
+						return reconcile.Result{}, err
+					}
+				}
 			}
-		}
 
-		err = r.maybeSetMultiStageAnnotation(pvc, datavolume)
-		if err != nil {
-			return reconcile.Result{}, err
+			err = r.maybeSetMultiStageAnnotation(pvc, datavolume)
+			if err != nil {
+				return reconcile.Result{}, err
+			}
 		}
-	}
 
-	switch selectedCloneStrategy {
-	case HostAssistedClone:
-		if err := r.ensureExtendedToken(pvc); err != nil {
-			return reconcile.Result{}, err
-		}
-	case CsiClone:
-		switch pvc.Status.Phase {
-		case corev1.ClaimBound:
-			if err := r.setCloneOfOnPvc(pvc); err != nil {
+		switch selectedCloneStrategy {
+		case HostAssistedClone:
+			if err := r.ensureExtendedToken(pvc); err != nil {
```

Comment:
I think this PR does a good job of handling the case where a DV has "pre populated" annotation and the PVC does not exist.  But it appears we have to better handle the case where the PVC has "populated for".

This bug has more details:  https://bugzilla.redhat.com/show_bug.cgi?id=2070366

Basically, the controller sees "populated for", takes ownership of the PVC, but still gets to this ` ensureExtendedToken` call

Reply:
do you want me to handle the problem with this PR, or create a new PR?

Reply:
updated the logic, now I need to prepare a test so this can be confirmed,

Reply:
looking at the tests